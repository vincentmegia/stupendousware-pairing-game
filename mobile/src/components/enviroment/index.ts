import {createContext} from 'react'

export const configuration = {environment: 'production'}
export const EnvironmentContext = createContext(configuration)
