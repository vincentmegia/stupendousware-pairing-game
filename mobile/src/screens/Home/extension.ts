import {createCard} from '@models/index'
import {CardModel} from '@models/types'

export const randomizer = (): string[] => {
  const list: string[] = []
  for (let index = 0; index < 6; index++) {
    list.push((Math.random() * 100).toFixed(0))
  }
  const cardsList = [...list, ...list]
  cardsList.sort(() => 0.5 - Math.random())
  return cardsList
}

export const getCards = (numbers: string[]): Array<CardModel[]> => {
  const list = numbers.map((number, index) => {
    return createCard(index, number)
  })
  //transform to row and cols
  const rows: Array<CardModel[]> = []
  let columns: CardModel[] = []
  list.forEach((item: CardModel, index: number) => {
    const tempIndex = ++index
    columns.push(item)
    if (tempIndex % 3 === 0) {
      rows.push(columns)
      columns = []
    }
  })
  return rows
}
