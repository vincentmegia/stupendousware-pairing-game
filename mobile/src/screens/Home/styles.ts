import {StyleSheet} from 'react-native'

export default StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'space-between',
    marginVertical: 16,
  },
  scoreBoard: {
    marginHorizontal: 16,
    marginTop: 16,
  },
  restartContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-between',
  },
  button: {
    fontSize: 16,
    color: 'blue',
  },
  text: {
    fontSize: 16,
  },
  textSmall: {
    fontSize: 12,
    marginTop: 16,
    color: 'grey',
  },
  debug: {marginLeft: 'auto', marginRight: 'auto', color: 'grey'},
})
