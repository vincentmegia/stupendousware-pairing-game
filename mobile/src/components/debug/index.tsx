import React, {useContext} from 'react'
import {Animated} from 'react-native'
import {EnvironmentContext} from '../enviroment'
import styles from './styles'

const Debug = () => {
  const {environment} = useContext(EnvironmentContext)
  return (
    <Animated.Text style={styles.text}>
      {environment === 'production' ? new Date().getMilliseconds() : 0}
    </Animated.Text>
  )
}

export default Debug
