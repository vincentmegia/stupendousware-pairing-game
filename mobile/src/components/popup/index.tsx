import React, {memo, ReactNode, useEffect, useState} from 'react'
import {Modal, View} from 'react-native'
import styles from './styles'

interface PopupProps {
  visible: boolean
  child: ReactNode
}
const Popup = ({visible, child}: PopupProps) => {
  const [isVisible, setVisible] = useState(visible)

  useEffect(() => {
    setVisible(visible)
  }, [visible])

  return (
    <View style={styles.container}>
      <Modal
        animationType="slide"
        transparent={true}
        visible={isVisible}
        onRequestClose={() => {
          setVisible(false)
        }}
        style={{flex: 1}}>
        {child}
      </Modal>
    </View>
  )
}

export default memo(Popup)
