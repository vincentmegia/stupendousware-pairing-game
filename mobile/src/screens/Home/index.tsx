import React, {useCallback, useRef, useState} from 'react'
import {Button, Pressable, SafeAreaView, Text, View} from 'react-native'
import CardList from '@components/card-list'
import Debug from '@components/debug'
import {CardModel} from '@models/types'
import {getCards, randomizer} from './extension'
import styles from './styles'
import Popup from '@components/popup'

const Home = () => {
  const cardsRef = useRef<Array<CardModel[]>>(getCards(randomizer()))
  const [pointerEvents, setPointerEvents] = useState<'none' | 'auto'>('auto')
  const [restart, setRestart] = useState(0)
  const [popup, setPopup] = useState(false)
  const [finished, setFinished] = useState(false)
  // for efficiency, since pointerEvents force re-render after every turn
  // will just use useRef for counting
  const totalMovesRef = useRef(0)
  const completedRef = useRef(false)

  const onRestart = useCallback(() => {
    setPopup(true)
  }, [])

  const onScoreUpdate = useCallback(
    (completed: boolean) => {
      if (completed) {
        setFinished(true)
        completedRef.current = true
        return
      }
      totalMovesRef.current = totalMovesRef.current + 1
    },
    [totalMovesRef],
  )

  const onConfirm = (confirm: boolean) => {
    setFinished(false)
    setPopup(false)
    if (!confirm) return
    completedRef.current = false
    cardsRef.current = getCards(randomizer())
    totalMovesRef.current = 0
    setRestart(new Date().getMilliseconds())
  }

  return (
    <SafeAreaView key={restart} style={styles.container}>
      <View style={styles.scoreBoard}>
        <View style={styles.restartContainer}>
          <Pressable onPress={onRestart}>
            <Text style={styles.button}>Restart</Text>
          </Pressable>
          <Text style={styles.text}>Total moves: {totalMovesRef.current}</Text>
        </View>
        <Text style={styles.textSmall}>
          The random numbers in the screen are re-render indicators. Application
          is optimized it doesnt re-render on card flip.
        </Text>
      </View>
      <View style={styles.container} pointerEvents={pointerEvents}>
        <CardList
          cards={cardsRef.current}
          onDisableClicks={() => {
            setPointerEvents('none')
            setTimeout(() => {
              setPointerEvents('auto')
            }, 1000)
          }}
          onScoreUpdate={onScoreUpdate}
        />
        <View style={styles.debug}>
          <Debug />
        </View>
      </View>

      <Popup
        visible={popup}
        child={
          <View style={styles.modal}>
            <Text style={styles.text}>
              {completedRef.current
                ? 'Do you wish to start new?'
                : 'Are you sure to restart in middle of game?'}
            </Text>
            <View style={styles.buttonGroup}>
              <Button title="No" onPress={() => onConfirm(false)} />
              <Button title="Yes" onPress={() => onConfirm(true)} />
            </View>
          </View>
        }
      />
      <Popup
        visible={finished}
        child={
          <View style={styles.modal}>
            <Text style={styles.text}>
              Congratulations! you've completed the game. Do you wish to play
              again?
            </Text>
            <View style={styles.buttonGroup}>
              <Button title="No" onPress={() => onConfirm(false)} />
              <Button title="Yes" onPress={() => onConfirm(true)} />
            </View>
          </View>
        }
      />
    </SafeAreaView>
  )
}

export default Home
