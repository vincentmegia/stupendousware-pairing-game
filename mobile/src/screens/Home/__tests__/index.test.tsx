import React from 'react'
import {render} from '@testing-library/react-native'
import Home from '..'
import {EnvironmentContext} from '../../../components/enviroment'

jest.mock('../extension', () => {
  return {
    randomizer: () => [],
    getCards: () => [
      [
        {
          id: 0,
          value: '1',
          matched: false,
          selected: false,
          reset: false,
        },
        {
          id: 1,
          value: '1',
          matched: false,
          selected: false,
          reset: false,
        },
        {
          id: 2,
          value: '2',
          matched: false,
          selected: false,
          reset: false,
        },
      ],
      [
        {
          id: 3,
          value: '2',
          matched: false,
          selected: false,
          reset: false,
        },
        {
          id: 4,
          value: '3',
          matched: false,
          selected: false,
          reset: false,
        },
        {
          id: 5,
          value: '3',
          matched: false,
          selected: false,
          reset: false,
        },
      ],
      [
        {
          id: 6,
          value: '4',
          matched: false,
          selected: false,
          reset: false,
        },
        {
          id: 7,
          value: '4',
          matched: false,
          selected: false,
          reset: false,
        },
        {
          id: 8,
          value: '5',
          matched: false,
          selected: false,
          reset: false,
        },
      ],
    ],
  }
})

describe('home screen tests', () => {
  it('it matches snapshots', () => {
    const {toJSON} = render(
      <EnvironmentContext.Provider value={{environment: 'test'}}>
        <Home />
      </EnvironmentContext.Provider>,
    )
    expect(toJSON()).toMatchSnapshot()
  })
})
