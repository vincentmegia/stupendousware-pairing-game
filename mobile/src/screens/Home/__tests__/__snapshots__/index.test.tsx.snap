// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`home screen tests it matches snapshots 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "justifyContent": "space-between",
      "marginVertical": 16,
    }
  }
>
  <View
    style={
      Object {
        "marginHorizontal": 16,
        "marginTop": 16,
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
      >
        <Text
          style={
            Object {
              "color": "blue",
              "fontSize": 16,
            }
          }
        >
          Restart
        </Text>
      </View>
      <Text
        style={
          Object {
            "fontSize": 16,
          }
        }
      >
        Total moves: 
        0
      </Text>
    </View>
    <Text
      style={
        Object {
          "color": "grey",
          "fontSize": 12,
          "marginTop": 16,
        }
      }
    >
      The random numbers in the screen are re-render indicators. Application is optimized it doesnt re-render on card flip.
    </Text>
  </View>
  <View
    pointerEvents="auto"
    style={
      Object {
        "flex": 1,
        "justifyContent": "space-between",
        "marginVertical": 16,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "elevation": 5,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.09,
            "shadowRadius": 4,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "180deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "fontWeight": "900",
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotateY": "180deg",
                  },
                ],
              }
            }
          >
            1
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backfaceVisibility": "hidden",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "0deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "backgroundColor": "white",
                "color": "white",
                "display": "none",
                "transform": Array [
                  Object {
                    "rotateY": "0deg",
                  },
                ],
              }
            }
          >
            1
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "elevation": 5,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.09,
            "shadowRadius": 4,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "180deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "fontWeight": "900",
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotateY": "180deg",
                  },
                ],
              }
            }
          >
            1
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backfaceVisibility": "hidden",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "0deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "backgroundColor": "white",
                "color": "white",
                "display": "none",
                "transform": Array [
                  Object {
                    "rotateY": "0deg",
                  },
                ],
              }
            }
          >
            1
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "elevation": 5,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.09,
            "shadowRadius": 4,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "180deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "fontWeight": "900",
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotateY": "180deg",
                  },
                ],
              }
            }
          >
            2
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backfaceVisibility": "hidden",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "0deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "backgroundColor": "white",
                "color": "white",
                "display": "none",
                "transform": Array [
                  Object {
                    "rotateY": "0deg",
                  },
                ],
              }
            }
          >
            2
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "elevation": 5,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.09,
            "shadowRadius": 4,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "180deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "fontWeight": "900",
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotateY": "180deg",
                  },
                ],
              }
            }
          >
            2
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backfaceVisibility": "hidden",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "0deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "backgroundColor": "white",
                "color": "white",
                "display": "none",
                "transform": Array [
                  Object {
                    "rotateY": "0deg",
                  },
                ],
              }
            }
          >
            2
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "elevation": 5,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.09,
            "shadowRadius": 4,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "180deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "fontWeight": "900",
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotateY": "180deg",
                  },
                ],
              }
            }
          >
            3
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backfaceVisibility": "hidden",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "0deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "backgroundColor": "white",
                "color": "white",
                "display": "none",
                "transform": Array [
                  Object {
                    "rotateY": "0deg",
                  },
                ],
              }
            }
          >
            3
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "elevation": 5,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.09,
            "shadowRadius": 4,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "180deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "fontWeight": "900",
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotateY": "180deg",
                  },
                ],
              }
            }
          >
            3
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backfaceVisibility": "hidden",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "0deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "backgroundColor": "white",
                "color": "white",
                "display": "none",
                "transform": Array [
                  Object {
                    "rotateY": "0deg",
                  },
                ],
              }
            }
          >
            3
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-around",
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "elevation": 5,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.09,
            "shadowRadius": 4,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "180deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "fontWeight": "900",
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotateY": "180deg",
                  },
                ],
              }
            }
          >
            4
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backfaceVisibility": "hidden",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "0deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "backgroundColor": "white",
                "color": "white",
                "display": "none",
                "transform": Array [
                  Object {
                    "rotateY": "0deg",
                  },
                ],
              }
            }
          >
            4
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "elevation": 5,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.09,
            "shadowRadius": 4,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "180deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "fontWeight": "900",
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotateY": "180deg",
                  },
                ],
              }
            }
          >
            4
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backfaceVisibility": "hidden",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "0deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "backgroundColor": "white",
                "color": "white",
                "display": "none",
                "transform": Array [
                  Object {
                    "rotateY": "0deg",
                  },
                ],
              }
            }
          >
            4
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "elevation": 5,
            "justifyContent": "center",
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.09,
            "shadowRadius": 4,
          }
        }
      >
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "180deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "fontWeight": "900",
                "position": "absolute",
                "transform": Array [
                  Object {
                    "rotateY": "180deg",
                  },
                ],
              }
            }
          >
            5
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
        <View
          collapsable={false}
          style={
            Object {
              "alignItems": "center",
              "backfaceVisibility": "hidden",
              "backgroundColor": "white",
              "borderRadius": 20,
              "justifyContent": "center",
              "maxHeight": 140,
              "maxWidth": 110,
              "minHeight": 130,
              "minWidth": 100,
              "position": "absolute",
              "transform": Array [
                Object {
                  "rotateY": "0deg",
                },
              ],
            }
          }
        >
          <Text
            collapsable={false}
            style={
              Object {
                "backfaceVisibility": "hidden",
                "backgroundColor": "white",
                "color": "white",
                "display": "none",
                "transform": Array [
                  Object {
                    "rotateY": "0deg",
                  },
                ],
              }
            }
          >
            5
          </Text>
          <Text
            collapsable={false}
            style={
              Object {
                "color": "grey",
                "fontSize": 10,
                "marginTop": 80,
              }
            }
          >
            0
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "color": "grey",
          "marginLeft": "auto",
          "marginRight": "auto",
        }
      }
    >
      <Text
        collapsable={false}
        style={
          Object {
            "color": "grey",
            "fontSize": 10,
            "marginTop": 80,
          }
        }
      >
        0
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "red",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      transparent={true}
      visible={false}
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "red",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
      transparent={true}
      visible={false}
    />
  </View>
</RCTSafeAreaView>
`;
